js/ui/iconGrid.js

```diff
@@ -770,6 +774,115 @@ const IconGrid = new Lang.Class({
+            let targetIndex = changedItems[sourceIndex];
+            let sourceActor = children[sourceIndex];
+            let actorOffset;
+
+            // There are two relevant cases:
+            //  1. We want to move icons between well-known positions. This usually
+            //     happens when the user moves an icon to inside a folder, and we
+            //     have to move the icons back.
+            //  2. We don't know the position. This happens when moving from inside
+            //     a folder to the main grid. The icons will be pushed forward by
+            //     one position, and the last icon will go to the unknown position.
+            if (targetIndex < children.length) {
+
+                // Case 1: we already have a previously allocated position. Just use that.
+                actorOffset = this._findActorOffset(sourceActor, children[targetIndex]);
+            } else {
+
+                // Case 2: We don't know where the icons should be moved to. Here,
+                // the solution is to create a temporary actor, allocate it, get
+                // the new position and then remove this temporary actor from the
+                // grid.
+                let stubWidget = new St.Widget({ visible: true,
+                                                 width: this._getHItemSize(),
+                                                 height: this._getVItemSize() });
+
+                this._grid.insert_child_at_index(stubWidget, -1);
+                this._allocate(this._grid, contentBox, 0);
+
+                // calculate the position of the new slot
+                actorOffset = this._findActorOffset(sourceActor, stubWidget);
+
+                // Remove the temporary actor from the grid
+                this._grid.remove_child(stubWidget);
+                this._allocate(this._grid, contentBox, 0);
+            }
+
+            movementMatrix[sourceIndex] = actorOffset;
+        }
+
+        // Make the original icon look like it fell into its place
+        let [originalIndex, dndDropPosition] = originalItemData;
+        let originalIcon = children[originalIndex];
+        if (originalIndex in movementMatrix) {
+            let oldIcon = children[originalIndex];
+            let newIcon = children[changedItems[originalIndex]];
+
+            // We need to know what the coordinates of the icon center are
+            dndDropPosition[0] -= Math.floor(oldIcon.get_size()[0] / 2);
+            dndDropPosition[1] -= Math.floor(oldIcon.get_size()[1] / 2);
+
+            // Draw it at the location where DnD accept occured
+            //let releaseOffset = this._findRelativeOffset(oldIcon, dndDropPosition);
+            //oldIcon.translation_x = releaseOffset[0];
+            //oldIcon.translation_y = releaseOffset[1];
+
+            movementMatrix[originalIndex] = this._findActorOffset(oldIcon, newIcon);
+        }
+
+        // Move icons that need animating
+        for (let sourceIndex in changedItems)
```

Comment:
I think this and the below loops also could benefit from using `for (let ... of ...)`

Reply:
Unless there's a way to assign 2 values from 2 variables, using the is index is still required, I think.

Reply:
True! I did not notice that the index was also used in the other variable.